examples/client-benchmark/main.cpp

```diff
@@ -0,0 +1,349 @@
+			pc = createPeerConnection(config, ws, id);
+		} else {
+			return;
+		}
+
+		if (type == "offer" || type == "answer") {
+			auto sdp = message["description"].get<string>();
+			pc->setRemoteDescription(Description(sdp, type));
+		} else if (type == "candidate") {
+			auto sdp = message["candidate"].get<string>();
+			auto mid = message["mid"].get<string>();
+			pc->addRemoteCandidate(Candidate(sdp, mid));
+		}
+	});
+
+	string wsPrefix = "";
+	if (params.webSocketServer().substr(0, 5).compare("ws://") != 0) {
+		wsPrefix = "ws://";
+	}
+	const string url = wsPrefix + params.webSocketServer() + ":" +
+	                   to_string(params.webSocketPort()) + "/" + localId;
+	cout << "Url is " << url << endl;
+	ws->open(url);
+
+	cout << "Waiting for signaling to be connected..." << endl;
+	wsFuture.get();
+
+	string id;
+	cout << "Enter a remote ID to send an offer:" << endl;
+	cin >> id;
+	cin.ignore();
+	if (id.empty()) {
+		// Nothing to do
+		return 0;
+	}
+	if (id == localId) {
+		cout << "Invalid remote ID (This is my local ID). Exiting..." << endl;
+		return 0;
+	}
+
+	cout << "Offering to " + id << endl;
+	auto pc = createPeerConnection(config, ws, id);
+
+	// We are the offerer, so create a data channel to initiate the process
+	const string label = "benchmark";
+	cout << "Creating DataChannel with label \"" << label << "\"" << endl;
+	auto dc = pc->createDataChannel(label);
+
+	dc->onOpen([id, wdc = make_weak_ptr(dc)]() {
+		cout << "DataChannel from " << id << " open" << endl;
+		if (noSend)
+			return;
+		if (auto dcLocked = wdc.lock()) {
+			cout << "Starting benchmark test. Sending data..." << endl;
+			try {
+				while (dcLocked->bufferedAmount() == 0) {
+					dcLocked->send(messageData);
+					sentSize += messageData.size();
+				}
+			} catch (const std::exception &e) {
+				std::cout << "Send failed: " << e.what() << std::endl;
+			}
+		}
+	});
+
+	dc->onBufferedAmountLow([wdc = make_weak_ptr(dc)]() {
+		if (noSend)
+			return;
+
+		auto dcLocked = wdc.lock();
+		if (!dcLocked)
+			return;
+
+		// Continue sending
+		try {
+			while (dcLocked->isOpen() && dcLocked->bufferedAmount() == 0) {
+				dcLocked->send(messageData);
+				sentSize += messageData.size();
+			}
+		} catch (const std::exception &e) {
+			std::cout << "Send failed: " << e.what() << std::endl;
+		}
+	});
+
+	dc->onClosed([id]() { cout << "DataChannel from " << id << " closed" << endl; });
+
+	dc->onMessage([id, wdc = make_weak_ptr(dc)](variant<binary, string> data) {
+		if (holds_alternative<binary>(data))
+			receivedSize += get<binary>(data).size();
+	});
+
+	dataChannelMap.emplace(id, dc);
+
+	const int duration = params.durationInSec() > 0 ? params.durationInSec() : INT32_MAX;
+	cout << "Becnhmark will run for " << duration << " seconds" << endl;
+
+	int printStatCounter = 0;
+
+	for (int i = 1; i <= duration; ++i) {
+		this_thread::sleep_for(1000ms);
+		cout << "#" << i << " Received: " << receivedSize.load() / 1024 << " KB/s"
```

Comment:
I would like a benchmarking client to properly compute throughputs. Here, the code relies on `sleep_for` being exact, which is not guaranteed, especially on a CPU under heavy load. Sleep calls are only guaranteed to wait at least for the requested duration, there is no guarantee on the upper bound. You should therefore compute the actual elapsed time like in`test/benchmark.cpp`.

Reply:
I know that but it is a not a promised constant throughput. It is just calculating approximate throughput.